models/user.go

```diff
@@ -1402,35 +1406,144 @@ func SyncExternalUsers() {
 					}
 
 					err = CreateUser(usr)
+
 					if err != nil {
 						log.Error(4, "SyncExternalUsers[%s]: Error creating user %s: %v", s.Name, su.Username, err)
+					} else if isAttributeSSHPublicKeySet {
+						// Add Public SSH Keys to User from LDAP
+						for _, LDAPPublicSSHKey := range su.SSHPublicKey {
+							log.Trace("SyncExternalUsers[%s]: Adding LDAP Public SSH Key for user %s: %v", s.Name, usr.Name, LDAPPublicSSHKey)
+							LDAPPublicSSHKeyName := strings.Join([]string{s.Name, LDAPPublicSSHKey[0:40]}, "-")
+							key, err := AddPublicKey(usr.ID, LDAPPublicSSHKeyName, LDAPPublicSSHKey, s.ID)
+							if err != nil {
+								log.Error(4, "SyncExternalUsers[%s]: Error adding LDAP Public SSH Key for user %s: %v [%v]", s.Name, su.Username, key, err)
+							} else {
+								sshKeysNeedUpdate = true
+							}
+						}
 					}
 				} else if updateExisting {
 					existingUsers = append(existingUsers, usr.ID)
-					// Check if user data has changed
-					if (len(s.LDAP().AdminFilter) > 0 && usr.IsAdmin != su.IsAdmin) ||
-						strings.ToLower(usr.Email) != strings.ToLower(su.Mail) ||
-						usr.FullName != fullName ||
-						!usr.IsActive {
-
-						log.Trace("SyncExternalUsers[%s]: Updating user %s", s.Name, usr.Name)
-
-						usr.FullName = fullName
-						usr.Email = su.Mail
-						// Change existing admin flag only if AdminFilter option is set
-						if len(s.LDAP().AdminFilter) > 0 {
-							usr.IsAdmin = su.IsAdmin
-						}
-						usr.IsActive = true
 
-						err = UpdateUser(usr)
+					if isAttributeSSHPublicKeySet {
+						log.Trace("SyncExternalUsers[%s]: Handling user %s", s.Name, usr.Name)
+
+						// Get Public Keys from DB with LDAP source
+						var giteaKeys []string
+						keys, err := ListPublicKeys(usr.ID)
+
 						if err != nil {
-							log.Error(4, "SyncExternalUsers[%s]: Error updating user %s: %v", s.Name, usr.Name, err)
+							log.Error(4, "SyncExternalUsers[%s]: Error listing LDAP Public SSH Keys for user %s: %v", s.Name, usr.Name, err)
+						}
+						for _, v := range keys {
+							// If key was synced from LDAP
+							if v.LoginSourceID > 0 {
+								giteaKeys = append(giteaKeys, v.OmitEmail())
+							}
+						}
+
+						// Get Public Keys from LDAP and skip duplicate keys
+						var ldapKeys []string
+						for _, v := range su.SSHPublicKey {
+							ldapKey := strings.Join(strings.Split(v, " ")[:2], " ")
+							// Check if value is already in slice, else add it
+							i := sort.SearchStrings(ldapKeys, ldapKey)
+							if i < len(ldapKeys) && ldapKeys[i] == ldapKey {
+							} else {
+								ldapKeys = append(ldapKeys, ldapKey)
+							}
+						}
+
+						// Sort key slices
+						sort.Strings(ldapKeys)
+						sort.Strings(giteaKeys)
+
+						// Check if sync is needed
+						var giteaKeysToDelete []string
+						if reflect.DeepEqual(giteaKeys, ldapKeys) {
```

Comment:
Because `reflect.DeepEqual` is so compute intense that you could just as well just sync all keys instead.

As for common interface, seems like `giteaKeys` and `ldapKeys` are of the same type no? In that case just make a compare function like so:
```go
func (pk *PublicKey) IsEqual(that *PublicKey) bool {
  // No need to check `Content`... if it has changed, so has `Fingerprint`...
  if that != nil &&
     that.Fingerprint == pk.Fingerprint &&
     that.Name == pk.Name &&
     that.Mode == pk.Mode &&
     that.Type == pk.Type {
    return true
  }
  return false
}